File: frontend/src/services/expense.js
import api from "./api";

export function normalizeExpense(expense) {
  return {
    ...expense,
    trip_id: expense.trip_id == null ? null : Number(expense.trip_id),
    vehicle_id: Number(expense.vehicle_id),
    amount: Number(expense.amount || 0),
    date: expense.date || expense.expense_date,
  };
}

export function toExpensePayload(values) {
  return {
    trip_id: values.trip_id ? Number(values.trip_id) : null,
    vehicle_id: Number(values.vehicle_id),
    type: values.type,
    amount: String(values.amount || 0),
    description: values.description || null,
    expense_date: values.expense_date || values.date,
  };
}

export async function listExpenses(params = {}) {
  const { data } = await api.get("/expenses/", { params });
  return data.map(normalizeExpense);
}

export async function createExpense(values) {
  const { data } = await api.post("/expenses/", toExpensePayload(values));
  return normalizeExpense(data);
}

export async function updateExpense(id, values) {
  const { data } = await api.put(`/expenses/${id}`, toExpensePayload(values));
  return normalizeExpense(data);
}

export async function deleteExpense(id) {
  await api.delete(`/expenses/${id}`);
}

File: backend/app/models/expense.py
from sqlalchemy import (
    Column,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    Numeric,
    Text
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base
from sqlalchemy import CheckConstraint, Index
from app.models.enums import ExpenseType


class Expense(Base):
    __tablename__ = "expenses"

    id = Column(Integer, primary_key=True, index=True)

    trip_id = Column(
        Integer,
        ForeignKey("trips.id"),
        nullable=False
    )

    vehicle_id = Column(
        Integer,
        ForeignKey("vehicles.id"),
        nullable=False
    )

    type = Column(
        Enum(ExpenseType),
        nullable=False
    )

    amount = Column(
        Numeric(12,2),
        nullable=False
    )

    description = Column(Text)

    expense_date = Column(
        Date,
        nullable=False
    )

    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now()
    )

    trip = relationship(
        "Trip",
        back_populates="expenses"
    )

    vehicle = relationship(
        "Vehicle",
        back_populates="expenses"
    )
    
    __table_args__ = (

        CheckConstraint(
            "amount >= 0",
            name="ck_expense_amount"
        ),

        Index("idx_expense_trip", "trip_id"),
        Index("idx_expense_vehicle", "vehicle_id"),
    )

    def __repr__(self):
        return f"<Expense {self.id}>"